server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1580,6 +1564,21 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
                 canMatchSearcher = searcher;
             } else {
                 indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
+                if (request.source() != null && (request.source().query() instanceof MatchNoneQueryBuilder) == false) {
+                    final SearchExecutionContext searchExecutionContext = new SearchExecutionContext(
+                        indexService.newSearchExecutionContext(
+                            request.shardId().getId(),
+                            request.shardRequestIndex(),
+                            null,
+                            request::nowInMillis,
+                            request.getClusterAlias(),
+                            request.getRuntimeMappings()
+                        )
+                    );
+                    if (queryStillMatchesAfterRewrite(request, searchExecutionContext) == false) {
+                        return new CanMatchShardResponse(false, null);
+                    }
+                }
```

Comment:
As I said in my previous comment, it's safe to do that for all cases (PIT and normal searcher). Please move the logic [here](https://github.com/elastic/elasticsearch/pull/96161/files/0de29feb016ae7516603a6198d427b03ebc35db6..1ce266bd43af65169d1d2b1cbf7def2a8230c12b#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fR1590).
Can you also add a comment explaining why rewriting without a searcher allows to ignore hasRefreshPending?